src/flamenco/vm/syscall/fd_vm_syscall_cpi_common.c

```diff
@@ -501,13 +695,44 @@ VM_SYSCALL_CPI_ENTRYPOINT( void *  _vm,
   vm->cu = vm->instr_ctx->txn_ctx->compute_meter;
 
   *_ret = instr_exec_res;
+
   if( FD_UNLIKELY( instr_exec_res ) ) return FD_VM_ERR_INSTR_ERR;
 
+  /* https://github.com/anza-xyz/agave/blob/b5f5c3cdd3f9a5859c49ebc27221dc27e143d760/programs/bpf_loader/src/syscalls/cpi.rs#L1128-L1145 */
+  /* Update all account permissions before updating the account data updates.
+     We have inlined the anza function update_caller_account_perms here.
+     TODO: consider factoring this out */
+  if( FD_FEATURE_ACTIVE( vm->instr_ctx->slot_ctx, bpf_account_data_direct_mapping ) ) {
+    for( ulong i=0UL; i<vm->instr_ctx->instr->acct_cnt; i++ ) {
+      /* Anza only even attemps to update the account permissions if it is a
+         "caller account". Only writable accounts are caller accounts. */
+      if( fd_instr_acc_is_writable_idx( vm->instr_ctx->instr, i ) ) {
+
+        uint is_writable = (uint)fd_account_can_data_be_changed( vm->instr_ctx->instr, i, &err );
+        /* Lookup memory regions for the account data and the realloc region. */
+        ulong data_region_idx    = vm->acc_region_metas[i].has_data_region ? vm->acc_region_metas[i].region_idx : 0;
+        ulong realloc_region_idx = vm->acc_region_metas[i].has_resizing_region ? vm->acc_region_metas[i].region_idx : 0;
+        if( data_region_idx && realloc_region_idx ) {
+          realloc_region_idx++;
+        }
+
+        if( data_region_idx ) {
+          vm->input_mem_regions[ data_region_idx ].is_writable = is_writable;
+        }
+        if( FD_LIKELY( realloc_region_idx ) ) { /* Unless is deprecated loader */
+          vm->input_mem_regions[ realloc_region_idx ].is_writable = is_writable;
+        }
+      }
+    }
+  }
+
   /* Update the caller accounts with any changes made by the callee during CPI execution */
   for( ulong i=0UL; i < caller_accounts_to_update_len; i++ ) {
-    fd_pubkey_t const * callee = &vm->instr_ctx->instr->acct_pubkeys[callee_account_keys[i]];
-    err = VM_SYSCALL_CPI_UPDATE_CALLER_ACC_FUNC(vm, &acc_infos[caller_accounts_to_update[i]], callee);
-    if( FD_UNLIKELY( err ) ) return err;
+    if( fd_instr_acc_is_writable_idx( vm->instr_ctx->instr, callee_account_keys[i] ) ) {
```

Comment:
Similarly, where is this logic in Agave?

Reply:
adding permalink. basically we only need to operate on writable accounts because non-writable accounts can't be updated